### DISCUSSION TITLE
Allow to write to task log from task policy

### DISCUSSION THREAD

**Discussion by alexInhert** [apache/airflow#18412] [General]
### Description

Airflow offers Task policies.
The problem is that while task policy alter the task and replace parameters there is no way to indicate to the user that parameters were switched.

### Use case/motivation

Consider the following:
```
def task_policy(task: BaseOperator):
    if task.timeout > timedelta(hours=10):
        task.timeout = timedelta(hours=10)
```

This replace the timeout for task if criteria met. However the problem is that if you set timeout 20 hours in your code this is what you will see in task instance details. You won't see 10 you will see 20.
Also if I do:
```
import logging
def task_policy(task: BaseOperator):
     log = logging.getLogger(__name__)
    if task.timeout > timedelta(hours=10):
        task.timeout = timedelta(hours=10)
        log.info("task policy changed timeout to 10 hours")
```

This log will be printed in the web server log! it will not be printed in the task log.

Result: the user who set timeout of 20 hours and the task timouted after 10 has no idea why it happened. He can't see it in the logs, he can't see it in task details. he is completely clueless.

Desired solution - allow to write to the task log from the task policy and if possible indicate in the task instance that a policy applied on the task and changed values.

### Related issues

no

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
As explained in #18402 I think in most cases you should not change values of the parameter but you should simply fail the Task if you see that it runs without the right parameter set. You should FAIL the task rather than mutate it in this case if it violates the Policy. There are some cases where you would likely want to mutate the task, but I think vast majority of cases is to fail it when the Policy is not met.

As opposed to "magic defaults" this has multiple good properties:

* it is very explicit
* it teaches people to use good practices (especially if your exceptions are informative, provide the user with explanation and guide them how to comply with the Policy.
* it allows you to make some more fine-grained check for the parameter values
 
I am converting that issue into discussion - I would like to hear why you would not be able to use it this way., I see that as non-issue, more of a choice that the Airflow project made.

**Reply by santosh-d3vpl3x** [reply]
In our setup, we have (ab)used `doc_md` field to highlight such (ab)use of cluster policies. Whenever a specific task param is modified, we also add a line in `doc_md` indicating the modification. This doc is visible in the airflow UI. Doing so, we sacrifice version control of such properties but for non-prod environments, it's largely a non-issue.
